src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -240,6 +213,104 @@ private AsyncProtocolRequest GetOrCreateProtocolRequest(ref AsyncProtocolRequest
+                    int payloadBytes = _sslState.GetRemainingFrameSize(_internalBuffer, _internalOffset, readBytes);
+                    if (payloadBytes < 0)
+                    {
+                        throw new IOException(SR.net_frame_read_size);
+                    }
+
+                    readBytes = await FillBufferAsync(adapter, SecureChannel.ReadHeaderSize + payloadBytes);
+                    if (readBytes < 0)
+                    {
+                        throw new IOException(SR.net_frame_read_size);
+                    }
+
+                    // At this point, readBytes contains the size of the header plus body.
+                    // Set _decrytpedBytesOffset/Count to the current frame we have (including header)
+                    // DecryptData will decrypt in-place and modify these to point to the actual decrypted data, which may be smaller.
+                    _decryptedBytesOffset = _internalOffset;
+                    _decryptedBytesCount = readBytes;
+                    SecurityStatusPal status = _sslState.DecryptData(_internalBuffer, ref _decryptedBytesOffset, ref _decryptedBytesCount);
+
+                    // Treat the bytes we just decrypted as consumed
+                    // Note, we won't do another buffer read until the decrypted bytes are processed
+                    ConsumeBufferedBytes(readBytes);
+
+                    if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
+                    {
+                        byte[] extraBuffer = null;
+                        if (_decryptedBytesCount != 0)
+                        {
+                            extraBuffer = new byte[_decryptedBytesCount];
+                            Buffer.BlockCopy(_internalBuffer, _decryptedBytesOffset, extraBuffer, 0, _decryptedBytesCount);
+
+                            _decryptedBytesCount = 0;
+                        }
+
+                        ProtocolToken message = new ProtocolToken(null, status);
+                        if (NetEventSource.IsEnabled)
+                            NetEventSource.Info(null, $"***Processing an error Status = {message.Status}");
+                        if (message.Renegotiate)
+                        {
+                            _sslState.ReplyOnReAuthentication(extraBuffer);
+
+                            // Loop on read.
+                            return -1;
+                        }
+
+                        if (message.CloseConnection)
+                        {
+                            _sslState.FinishRead(null);
+                            return 0;
+                        }
+
+                        throw new IOException(SR.net_io_decrypt, message.GetException());
+                    }
+                }
+                catch (Exception e)
+                {
+                    if (e is IOException)
+                    {
+                        throw;
+                    }
+
```

Comment:
Could be "catch (Exception e) when (!(e is IOException))"

Reply:
done